File: app/src/main/java/io/github/anshu7vyas/stocked/StockedApp.java
package io.github.anshu7vyas.stocked;


import android.app.Application;
import android.app.NotificationChannel;
import android.app.NotificationManager;

/**
 * Class represents the whole application. Owns app-wide setup such as the notification channel.
 */
public class StockedApp extends Application {

    public static final String CHANNEL_EXPIRY = "expiry_alerts";

    @Override
    public void onCreate() {
        super.onCreate();
        NotificationChannel channel = new NotificationChannel(
                CHANNEL_EXPIRY,
                getString(R.string.channel_expiry_name),
                NotificationManager.IMPORTANCE_DEFAULT);
        channel.setDescription(getString(R.string.channel_expiry_description));
        getSystemService(NotificationManager.class).createNotificationChannel(channel);
    }
}

File: app/src/main/java/io/github/anshu7vyas/stocked/StockedApp.kt
package io.github.anshu7vyas.stocked

import android.app.Application
import android.app.NotificationChannel
import android.app.NotificationManager
import androidx.hilt.work.HiltWorkerFactory
import androidx.work.Configuration
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import dagger.hilt.android.HiltAndroidApp
import io.github.anshu7vyas.stocked.work.ExpiryCheckWorker
import java.util.concurrent.TimeUnit
import javax.inject.Inject

@HiltAndroidApp
class StockedApp : Application(), Configuration.Provider {

    @Inject lateinit var workerFactory: HiltWorkerFactory

    override val workManagerConfiguration: Configuration
        get() = Configuration.Builder().setWorkerFactory(workerFactory).build()

    override fun onCreate() {
        super.onCreate()
        createNotificationChannel()
        scheduleExpiryCheck()
    }

    private fun createNotificationChannel() {
        val channel = NotificationChannel(
            CHANNEL_EXPIRY,
            getString(R.string.channel_expiry_name),
            NotificationManager.IMPORTANCE_DEFAULT,
        ).apply {
            description = getString(R.string.channel_expiry_description)
        }
        getSystemService(NotificationManager::class.java).createNotificationChannel(channel)
    }

    private fun scheduleExpiryCheck() {
        WorkManager.getInstance(this).enqueueUniquePeriodicWork(
            ExpiryCheckWorker.WORK_NAME,
            ExistingPeriodicWorkPolicy.KEEP,
            PeriodicWorkRequestBuilder<ExpiryCheckWorker>(1, TimeUnit.DAYS).build(),
        )
    }

    companion object {
        const val CHANNEL_EXPIRY = "expiry_alerts"
    }
}
